snuba/utils/streams/strategy_factory.py

```diff
@@ -0,0 +1,102 @@
+from abc import abstractmethod
+from typing import Callable, Mapping, Optional, Protocol, TypeVar
+
+from snuba.utils.streams.backends.kafka import KafkaPayload
+from snuba.utils.streams.metrics import DummyMetricsBackend, Metrics, MetricsWrapper
+from snuba.utils.streams.processing.strategies import (
+    ProcessingStrategy,
+    ProcessingStrategyFactory,
+)
+from snuba.utils.streams.processing.strategies.streaming import (
+    CollectStep,
+    FilterStep,
+    ParallelTransformStep,
+    TransformStep,
+)
+from snuba.utils.streams.types import Message, Partition
+
+
+class StreamMessageFilter(Protocol):
+    """
+    A filter over messages coming from a stream. Can be used to pre filter
+    messages during consumption but potentially for other use cases as well.
+    """
+
+    @abstractmethod
+    def should_drop(self, message: Message[KafkaPayload]) -> bool:
+        raise NotImplementedError
+
+
+TProcessed = TypeVar("TProcessed")
+
+
+class StreamingConsumerStrategyFactory(ProcessingStrategyFactory[KafkaPayload]):
+    def __init__(
+        self,
+        prefilter: Optional[StreamMessageFilter],
+        process_message: Callable[[Message[KafkaPayload]], TProcessed],
+        collector: Callable[[], ProcessingStrategy[TProcessed]],
```

Comment:
Good thanks. I like this.

Can we parametrize the Payload instead of using `KafkaPayload` ? This abstraction seems not to be Kafka specific now.

Reply:
I changed this class to support parametrized payload. I also added `KafkaConsumerStrategyFactory(ConsumerStrategyFactory[KafkaPayload])` though which is the one I'd expect end users to mostly use.